File: projects/android/Sample/app/src/main/java/com/ezored/sample/ui/fragment/base/BaseFragment.java
package com.ezored.sample.ui.fragment.base;

import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.ezored.sample.R;
import com.ezored.sample.enums.LoadStateEnum;
import com.ezored.sample.ui.activity.base.BaseActivity;
import com.ezored.sample.utils.DateTimeUtils;
import com.ezored.sample.utils.UIUtil;
import com.ezored.util.Logger;

import org.greenrobot.eventbus.EventBus;

public class BaseFragment extends Fragment {

    protected LoadStateEnum remoteDataLoadState = LoadStateEnum.NOT_LOADED;
    protected long lastCacheUpdate = 0;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Bundle arguments = getArguments();

        if (arguments != null) {
            onGetArguments(arguments);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(getFragmentLayout(), container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        createAll(view);
        layoutAll(view);

        showMainView();
    }

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        super.setUserVisibleHint(isVisibleToUser);

        if (isVisibleToUser) {
            onFragmentVisible();
        } else {
            onFragmentNotVisible();
        }
    }

    @Override
    public void onStart() {
        super.onStart();

        if (registerForEventBus()) {
            EventBus.getDefault().register(this);
        }
    }

    @Override
    public void onStop() {
        super.onStop();

        if (registerForEventBus()) {
            EventBus.getDefault().unregister(this);
        }
    }

    protected void hideAllViews() {
        UIUtil.hideViewById(getView(), R.id.main_view);
        UIUtil.hideViewById(getView(), R.id.network_error_view);
        UIUtil.hideViewById(getView(), R.id.error_view);
        UIUtil.hideViewById(getView(), R.id.loading_view);
    }

    protected void showMainView() {
        hideAllViews();

        if (getView() != null && getView().findViewById(R.id.main_view) != null) {
            getView().findViewById(R.id.main_view).setBackgroundColor(ContextCompat.getColor(getContext(), getViewBackgroundColor()));
            UIUtil.showViewById(getView(), R.id.main_view);
        }
    }

    protected void showLoadingView() {
        hideAllViews();

        if (getView() != null && getView().findViewById(R.id.loading_view) != null) {
            getView().findViewById(R.id.loading_view).setBackgroundColor(ContextCompat.getColor(getContext(), getViewBackgroundColor()));

            ProgressBar progressBar = (ProgressBar) getView().findViewById(R.id.pbLoadingView);
            progressBar.getIndeterminateDrawable().setColorFilter(ContextCompat.getColor(getContext(), getLoadingViewProgressBarColor()), PorterDuff.Mode.MULTIPLY);

            UIUtil.showViewById(getView(), R.id.loading_view);
        }
    }

    protected void showNetworkView(String message) {
        hideAllViews();

        if (getView() != null && getView().findViewById(R.id.network_error_view) != null) {
            getView().findViewById(R.id.network_error_view).setBackgroundColor(ContextCompat.getColor(getContext(), getViewBackgroundColor()));

            TextView tvMessage = (TextView) getView().findViewById(R.id.tv_message);
            tvMessage.setText(message);

            Button btNetworkError = (Button) getView().findViewById(R.id.bt_network_error);

            btNetworkError.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    onBtNetworkErrorClick();
                }
            });

            UIUtil.showViewById(getView(), R.id.network_error_view);
        }
    }

    protected void showErrorView(String message) {
        hideAllViews();

        if (getView() != null && getView().findViewById(R.id.error_view) != null) {
            getView().findViewById(R.id.error_view).setBackgroundColor(ContextCompat.getColor(getContext(), getViewBackgroundColor()));

            TextView tvMessage = (TextView) getView().findViewById(R.id.tv_message);
            tvMessage.setText(message);

            UIUtil.showViewById(getView(), R.id.error_view);
        }
    }

    protected int getViewBackgroundColor() {
        return R.color.white;
    }

    protected int getLoadingViewProgressBarColor() {
        return R.color.color_accent;
    }

    protected void onBtErrorClick() {
        // ignore
    }

    protected void onBtNetworkErrorClick() {
        // ignore
    }

    protected void onFragmentNotVisible() {
        // ignore
    }

    protected void onFragmentVisible() {
        if (getActivity() != null) {
            // notify all analytics about it
            // AppEvents.onScreenChange(getScreenNameForAnalytics(), getActivity());
        }

        validateLoadData();
    }

    protected int getFragmentLayout() {
        return 0;
    }

    protected void onGetArguments(Bundle arguments) {

    }

    protected void createAll(View view) {
        view.setBackgroundColor(ContextCompat.getColor(getContext(), getViewBackgroundColor()));
    }

    protected void layoutAll(View view) {
        // ignore
    }

    protected boolean registerForEventBus() {
        return false;
    }

    protected String getScreenNameForAnalytics() {
        return null;
    }

    /////////////////////////////////////////////
    // REMOTE DATA
    /////////////////////////////////////////////

    protected void loadData() {
        if (!needLoadNewData()) {
            return;
        }

        if (remoteDataLoadState != LoadStateEnum.NOT_LOADED) {
            return;
        }

        remoteDataLoadState = LoadStateEnum.LOADING;

        onLoadNewData();
    }

    protected void validateLoadData() {
        validateCache();
        loadData();
    }

    protected void onLoadNewData() {
        // ignore
    }

    protected boolean needLoadNewData() {
        return false;
    }

    protected boolean isLoading() {
        return remoteDataLoadState == LoadStateEnum.LOADING;
    }

    /////////////////////////////////////////////
    // CACHE
    /////////////////////////////////////////////

    protected void onCacheExpired() {
        Logger.i("[BaseFragment : onCacheExpired] Cache expired now");
    }

    protected boolean hasCache() {
        return false;
    }

    protected long cacheExpireInterval() {
        return 0;
    }

    protected void validateCache() {
        if (hasCache()) {
            lastCacheUpdate = lastCacheUpdate > 0 ? lastCacheUpdate : DateTimeUtils.getCurrentTimestamp();

            long currentTime = DateTimeUtils.getCurrentTimestamp();
            long expireTime = lastCacheUpdate + cacheExpireInterval();

            if (currentTime >= expireTime) {
                lastCacheUpdate = DateTimeUtils.getCurrentTimestamp();
                onCacheExpired();
            }
        }
    }

    /////////////////////////////////////////////
    // TOOLBAR
    /////////////////////////////////////////////

    public void setupToolbar(int resId) {
        setupToolbar(getString(resId));
    }

    public void setupToolbar(String title) {
        View view = getView();

        if (view != null) {
            Toolbar toolbar = (Toolbar) view.findViewById(R.id.toolbar);

            if (toolbar != null) {
                toolbar.setTitle("");
                toolbar.setBackgroundColor(ContextCompat.getColor(getContext(), getToolbarBackgroundColor()));
                toolbar.setTitleTextColor(ContextCompat.getColor(getContext(), getToolbarTextColor()));

                ((BaseActivity) getActivity()).setSupportActionBar(toolbar);
                getSupportActionBar().setElevation(UIUtil.convertDpToPixel(4));

                final Drawable upArrow = ContextCompat.getDrawable(getContext(), R.drawable.ic_arrow_back);
                upArrow.setColorFilter(ContextCompat.getColor(getContext(), getToolbarIconColor()), PorterDuff.Mode.SRC_ATOP);
                getSupportActionBar().setHomeAsUpIndicator(upArrow);

                if (hasToolbarLogo()) {
                    ImageView imToolbarLogo = (ImageView) view.findViewById(R.id.iv_toolbar_logo);

                    if (imToolbarLogo != null) {
                        onSetToolbarLogo(imToolbarLogo);
                    }
                } else {
                    if (title != null) {
                        TextView tvToolbarTitle = (TextView) view.findViewById(R.id.tv_toolbar_title);

                        if (tvToolbarTitle != null) {
                            tvToolbarTitle.setText(title);
                            tvToolbarTitle.setTextColor(ContextCompat.getColor(getContext(), getToolbarTextColor()));
                        }
                    }
                }
            }
        }
    }

    public void onSetToolbarLogo(ImageView logo) {
        if (getContext() != null) {
            if (getPaintToolbarLogo()) {
                logo.setImageDrawable(UIUtil.drawableColorChange(getContext(), getToolbarLogo(), getToolbarTextColor()));
            } else {
                logo.setImageDrawable(ContextCompat.getDrawable(getContext(), getToolbarLogo()));
            }
        }
    }

    private boolean getPaintToolbarLogo() {
        return true;
    }

    protected Toolbar getToolbar() {
        View view = getView();

        if (view != null) {
            return (Toolbar) view.findViewById(R.id.toolbar);
        }

        return null;
    }

    protected int getToolbarLogo() {
        return R.drawable.ic_toolbar_logo;
    }

    protected int getToolbarIconColor() {
        return R.color.white;
    }

    protected int getToolbarBackgroundColor() {
        return R.color.color_primary;
    }

    protected int getToolbarTextColor() {
        return R.color.white;
    }

    protected void showToolbarBackButton(boolean show) {
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(show);
            getSupportActionBar().setDisplayShowHomeEnabled(show);
        }
    }

    protected boolean hasToolbarLogo() {
        return false;
    }

    public ActionBar getSupportActionBar() {
        if (getActivity() != null) {
            return ((BaseActivity) getActivity()).getSupportActionBar();
        }

        return null;
    }

}

File: projects/android/Sample/app/src/main/java/com/ezored/sample/ui/fragment/base/BaseFragment.kt
package com.ezored.sample.ui.fragment.base

import android.graphics.PorterDuff
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.appcompat.app.ActionBar
import androidx.appcompat.widget.Toolbar
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import com.ezored.sample.R
import com.ezored.sample.enumerator.LoadStateEnum
import com.ezored.sample.ui.activity.base.BaseActivity
import com.ezored.sample.util.DateTimeUtil
import com.ezored.sample.util.UIUtil
import com.ezored.util.Logger
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import org.greenrobot.eventbus.EventBus

open class BaseFragment : Fragment(), CoroutineScope {

    override val coroutineContext = Dispatchers.Main

    protected var remoteDataLoadState = LoadStateEnum.NOT_LOADED

    protected var lastCacheUpdate: Long = 0

    protected val viewBackgroundColor: Int
        get() = R.color.white

    protected val loadingViewProgressBarColor: Int
        get() = R.color.color_accent

    protected open val fragmentLayout: Int
        get() = 0

    protected open val screenNameForAnalytics: String?
        get() = null

    protected val isLoading: Boolean
        get() = remoteDataLoadState == LoadStateEnum.LOADING

    private val paintToolbarLogo: Boolean
        get() = true

    protected val toolbar: Toolbar?
        get() = view?.findViewById(R.id.toolbar) as Toolbar

    protected val toolbarLogo: Int
        get() = R.drawable.ic_toolbar_logo

    protected val toolbarIconColor: Int
        get() = R.color.white

    protected val toolbarBackgroundColor: Int
        get() = R.color.color_primary

    protected val toolbarTextColor: Int
        get() = R.color.white

    protected val supportActionBar: ActionBar?
        get() = (activity as BaseActivity).supportActionBar

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val arguments = arguments

        if (arguments != null) {
            onGetArguments(arguments)
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(fragmentLayout, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        createAll(view)
        layoutAll(view)

        showMainView()
    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)

        if (isVisibleToUser) {
            onFragmentVisible()
        } else {
            onFragmentNotVisible()
        }
    }

    override fun onStart() {
        super.onStart()

        if (registerForEventBus()) {
            EventBus.getDefault().register(this)
        }
    }

    override fun onStop() {
        super.onStop()

        if (registerForEventBus()) {
            EventBus.getDefault().unregister(this)
        }
    }

    protected fun hideAllViews() {
        UIUtil.hideViewById(view, R.id.main_view)
        UIUtil.hideViewById(view, R.id.network_error_view)
        UIUtil.hideViewById(view, R.id.error_view)
        UIUtil.hideViewById(view, R.id.loading_view)
    }

    protected fun showMainView() {
        hideAllViews()

        context?.let { context ->
            view?.let { view ->
                view.findViewById<View>(R.id.main_view)?.let { mainView ->
                    mainView.setBackgroundColor(
                        ContextCompat.getColor(
                            context,
                            viewBackgroundColor
                        )
                    )
                }

                UIUtil.showViewById(view, R.id.main_view)
            }
        }
    }

    protected fun showLoadingView() {
        hideAllViews()

        context?.let { context ->
            view?.let { view ->
                view.findViewById<View>(R.id.loading_view)?.let { loadingView ->
                    loadingView.setBackgroundColor(
                        ContextCompat.getColor(
                            context,
                            viewBackgroundColor
                        )
                    )
                }

                view.findViewById<ProgressBar>(R.id.pbLoadingView)?.let { pbLoadingView ->
                    pbLoadingView.indeterminateDrawable.setColorFilter(
                        ContextCompat.getColor(
                            context,
                            loadingViewProgressBarColor
                        ),
                        PorterDuff.Mode.MULTIPLY
                    )
                }

                UIUtil.showViewById(view, R.id.loading_view)
            }
        }
    }

    protected fun showNetworkView(message: String) {
        hideAllViews()

        context?.let { context ->
            view?.let { view ->
                view.findViewById<View>(R.id.network_error_view)?.let { networkErrorView ->
                    networkErrorView.setBackgroundColor(
                        ContextCompat.getColor(
                            context,
                            viewBackgroundColor
                        )
                    )

                    networkErrorView.findViewById<TextView>(R.id.tv_message)?.let { tvMessage ->
                        tvMessage.text = message
                    }

                    view.findViewById<Button>(R.id.bt_network_error)?.let { btNetworkError ->
                        btNetworkError.setOnClickListener { onBtNetworkErrorClick() }
                    }
                }

                UIUtil.showViewById(view, R.id.network_error_view)
            }
        }
    }

    protected fun showErrorView(message: String) {
        hideAllViews()

        context?.let { context ->
            view?.let { view ->
                view.findViewById<View>(R.id.error_view)?.let { errorView ->
                    errorView.setBackgroundColor(
                        ContextCompat.getColor(
                            context,
                            viewBackgroundColor
                        )
                    )

                    errorView.findViewById<TextView>(R.id.tv_message)?.let { tvMessage ->
                        tvMessage.text = message
                    }
                }

                UIUtil.showViewById(view, R.id.error_view)
            }
        }
    }

    protected fun onBtNetworkErrorClick() {
        // ignore
    }

    protected fun onFragmentNotVisible() {
        // ignore
    }

    protected fun onFragmentVisible() {
        activity?.let { activity ->
            // notify all analytics about it
            // AppEvents.onScreenChange(getScreenNameForAnalytics(), activity);
        }

        validateLoadData()
    }

    protected fun onGetArguments(arguments: Bundle) {
        // ignore
    }

    protected open fun createAll(view: View) {
        context?.let { context ->
            view.setBackgroundColor(ContextCompat.getColor(context, viewBackgroundColor))
        }
    }

    protected fun layoutAll(view: View) {
        // ignore
    }

    protected fun registerForEventBus(): Boolean {
        return false
    }

    // ///////////////////////////////////////////
    // REMOTE DATA
    // ///////////////////////////////////////////

    protected fun loadData() {
        if (!needLoadNewData()) {
            return
        }

        if (remoteDataLoadState != LoadStateEnum.NOT_LOADED) {
            return
        }

        remoteDataLoadState = LoadStateEnum.LOADING

        onLoadNewData()
    }

    protected fun validateLoadData() {
        validateCache()
        loadData()
    }

    protected open fun onLoadNewData() {
        // ignore
    }

    protected open fun needLoadNewData(): Boolean {
        return false
    }

    // ///////////////////////////////////////////
    // CACHE
    // ///////////////////////////////////////////

    protected fun onCacheExpired() {
        Logger.i("[BaseFragment : onCacheExpired] Cache expired now")
    }

    protected fun hasCache(): Boolean {
        return false
    }

    protected fun cacheExpireInterval(): Long {
        return 0
    }

    protected fun validateCache() {
        if (hasCache()) {
            lastCacheUpdate =
                if (lastCacheUpdate > 0) lastCacheUpdate else DateTimeUtil.currentTimestamp

            val currentTime = DateTimeUtil.currentTimestamp
            val expireTime = lastCacheUpdate + cacheExpireInterval()

            if (currentTime >= expireTime) {
                lastCacheUpdate = DateTimeUtil.currentTimestamp
                onCacheExpired()
            }
        }
    }

    // ///////////////////////////////////////////
    // TOOLBAR
    // ///////////////////////////////////////////

    fun setupToolbar(resId: Int) {
        setupToolbar(getString(resId))
    }

    fun setupToolbar(title: String?) {
        context?.let { context ->
            view?.let { view ->
                view.findViewById<Toolbar>(R.id.toolbar)?.let { toolbar ->
                    toolbar.title = ""
                    toolbar.setBackgroundColor(
                        ContextCompat.getColor(
                            context,
                            toolbarBackgroundColor
                        )
                    )
                    toolbar.setTitleTextColor(ContextCompat.getColor(context, toolbarTextColor))

                    (activity as BaseActivity).setSupportActionBar(toolbar)

                    supportActionBar?.let { supportActionBar ->
                        supportActionBar.elevation = UIUtil.convertDpToPixel(4f)

                        ContextCompat.getDrawable(context, R.drawable.ic_arrow_back)
                            ?.let { backArrowDrawable ->
                                backArrowDrawable.setColorFilter(
                                    ContextCompat.getColor(context, toolbarIconColor),
                                    PorterDuff.Mode.SRC_ATOP
                                )

                                supportActionBar.setHomeAsUpIndicator(backArrowDrawable)
                            }
                    }

                    if (hasToolbarLogo()) {
                        view.findViewById<ImageView>(R.id.iv_toolbar_logo)?.let { imToolbarLogo ->
                            onSetToolbarLogo(imToolbarLogo)
                        }
                    } else {
                        title?.let { title ->
                            view.findViewById<TextView>(R.id.tv_toolbar_title)
                                ?.let { tvToolbarTitle ->
                                    tvToolbarTitle.text = title
                                    tvToolbarTitle.setTextColor(
                                        ContextCompat.getColor(
                                            context,
                                            toolbarTextColor
                                        )
                                    )
                                }
                        }
                    }
                }
            }
        }
    }

    fun onSetToolbarLogo(logo: ImageView) {
        context?.let { context ->
            if (paintToolbarLogo) {
                logo.setImageDrawable(
                    UIUtil.drawableColorChange(
                        context,
                        toolbarLogo,
                        toolbarTextColor
                    )
                )
            } else {
                logo.setImageDrawable(ContextCompat.getDrawable(context, toolbarLogo))
            }
        }
    }

    protected fun showToolbarBackButton(show: Boolean) {
        supportActionBar?.setDisplayHomeAsUpEnabled(show)
        supportActionBar?.setDisplayShowHomeEnabled(show)
    }

    protected fun hasToolbarLogo(): Boolean {
        return false
    }
}
